crates/turbo-tasks/src/task/function.rs

```diff
@@ -155,50 +155,20 @@ macro_rules! task_inputs_impl {
     }
 }
 
-macro_rules! as_concrete_task_input {
-    ( $arg:ident ) => {
-        ConcreteTaskInput
-    };
-}
-
 macro_rules! task_fn_impl {
-    ( $helper_module:ident $async_fn_trait:ident $arg_len:literal $( $arg:ident )* ) => {
-        mod $helper_module {
-            use super::*;
-
-            // this is a non-generic helper method to allow code-sharing across monomorphized
-            // instances of this function
-            pub fn get_args(
-                inputs: &[ConcreteTaskInput],
-            ) -> Result<($(&as_concrete_task_input!($arg),)*)> {
-                get_args_iter(inputs.iter())
-            }
-
-            fn get_args_iter(
-                mut iter: std::slice::Iter<'_, ConcreteTaskInput>,
-            ) -> Result<($(&as_concrete_task_input!($arg),)*)> {
-                let args = ($(next_arg(&mut iter, stringify!($arg))?,)*);
-                if iter.next().is_some() {
-                    bail!("task was called with too many arguments");
-                }
-                Ok(args)
-            }
-        }
-
+    ( $async_fn_trait:ident $arg_len:literal $( $arg:ident )* ) => {
         impl<F, Output, $($arg,)*> TaskFnInputFunction<FunctionMode, ($($arg,)*)> for F
         where
             $($arg: TaskInput + 'static,)*
             F: Fn($($arg,)*) -> Output + Send + Sync + Clone + 'static,
             Output: TaskOutput + 'static,
         {
             #[allow(non_snake_case)]
-            fn functor(&self, _this: Option<RawVc>, inputs: &[ConcreteTaskInput]) -> Result<NativeTaskFn> {
-                let ($($arg,)*) = $helper_module::get_args(inputs)?;
+            fn functor(&self, _this: Option<RawVc>, arg: &ConcreteTaskInput) -> Result<NativeTaskFn> {
+                #[allow(unused_parens)]
+                let ($($arg),*) = <($($arg),*) as TaskInput>::try_from_concrete(arg)?;
```

Comment:
Nit: Always creating a tuple by always inserting a trailing comma removes the need for ignoring the lint.

```suggestion
                let ($($arg,)*) = <($($arg,)*) as TaskInput>::try_from_concrete(arg)?;
```